Some way to create a const rational value?

I'd like to create a `const` Rational64 value (for use in a trait impl). 

The `Ratio` struct fields are private and the various constructors can't be used in a `const` expression. Is there some other way of doing this?

It appears as if `const fn` is stable as of [Rust 1.13.0](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1310-2018-12-06). That being said, to take a good first stab at  arbitrary `const Ratio<T>`s, we need some way to verify that the denominator is not zero with another `const fn`, i.e. we can't do this
```
pub const fn new_const(numer: T, denom: T) -> Ratio<T> {
    if denom.is_zero() {
        ...
```
because is_zero() is not a `const fn` and `const fn`s can call only other `const fn`s. As the compiler puts it
```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
   |
   |         if denom.is_zero() {
   |            ^^^^^^^^^^^^^^^
```
We would have to add a `const fn` version of `is_zero` in the [num-traits](https://github.com/rust-num/num-traits) crate. For consistency, we would want to do so for a variety of other functions as well (including `gcd()`, also from num-traits, used in `reduce()`).

Before implementing arbitrary const Ratios, there should be a larger plan for consts in the [rust-num](https://github.com/rust-num/num) crates.

--- a/build.rs
+++ b/build.rs
@@ -4,10 +4,16 @@ use std::env;
 
 fn main() {
     let ac = autocfg::new();
+
     if ac.probe_type("i128") {
         println!("cargo:rustc-cfg=has_i128");
     } else if env::var_os("CARGO_FEATURE_I128").is_some() {
         panic!("i128 support was not detected!");
+    }
+
+    // autocfg doesn't have a direct way to probe for `const fn` yet.
+    if ac.probe_rustc_version(1, 31) {
+        autocfg::emit("has_const_fn");
     }
 
     autocfg::rerun_path(file!());

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -68,6 +68,42 @@ pub type Rational64 = Ratio<i64>;
 /// Alias for arbitrary precision rationals.
 pub type BigRational = Ratio<BigInt>;
 
+macro_rules! maybe_const {
+    ($( $(#[$attr:meta])* pub fn $name:ident $args:tt -> $ret:ty $body:block )*) => {$(
+        #[cfg(has_const_fn)]
+        $(#[$attr])* pub const fn $name $args -> $ret $body
+
+        #[cfg(not(has_const_fn))]
+        $(#[$attr])* pub fn $name $args -> $ret $body
+    )*}
+}
+
+/// These method are `const` for Rust 1.31 and later.
+impl<T> Ratio<T> {
+    maybe_const! {
+        /// Creates a `Ratio` without checking for `denom == 0` or reducing.
+        #[inline]
+        pub fn new_raw(numer: T, denom: T) -> Ratio<T> {
+            Ratio {
+                numer: numer,
+                denom: denom,
+            }
+        }
+
+        /// Gets an immutable reference to the numerator.
+        #[inline]
+        pub fn numer(&self) -> &T {
+            &self.numer
+        }
+
+        /// Gets an immutable reference to the denominator.
+        #[inline]
+        pub fn denom(&self) -> &T {
+            &self.denom
+        }
+    }
+}
+
 impl<T: Clone + Integer> Ratio<T> {
     /// Creates a new `Ratio`. Fails if `denom` is zero.
     #[inline]
@@ -86,31 +122,10 @@ impl<T: Clone + Integer> Ratio<T> {
         Ratio::new_raw(t, One::one())
     }
 
-    /// Creates a `Ratio` without checking for `denom == 0` or reducing.
-    #[inline]
-    pub fn new_raw(numer: T, denom: T) -> Ratio<T> {
-        Ratio {
-            numer: numer,
-            denom: denom,
-        }
-    }
-
     /// Converts to an integer, rounding towards zero.
     #[inline]
     pub fn to_integer(&self) -> T {
         self.trunc().numer
-    }
-
-    /// Gets an immutable reference to the numerator.
-    #[inline]
-    pub fn numer<'a>(&'a self) -> &'a T {
-        &self.numer
-    }
-
-    /// Gets an immutable reference to the denominator.
-    #[inline]
-    pub fn denom<'a>(&'a self) -> &'a T {
-        &self.denom
     }
 
     /// Returns true if the rational number is an integer (denominator is 1).
@@ -2070,5 +2085,20 @@ mod test {
 
         r.set_one();
         assert!(r.is_one());
+    }
+
+    #[cfg(has_const_fn)]
+    #[test]
+    fn test_const() {
+        const N: Ratio<i32> = Ratio::new_raw(123, 456);
+        const N_NUMER: &i32 = N.numer();
+        const N_DENOM: &i32 = N.denom();
+
+        assert_eq!(N_NUMER, &123);
+        assert_eq!(N_DENOM, &456);
+
+        let r = N.reduced();
+        assert_eq!(r.numer(), &(123 / 3));
+        assert_eq!(r.denom(), &(456 / 3));
     }
 }

